File: iosApp/iosApp/Modules/Authenticate/AuthenticatePresenter.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

enum AuthenticatePresenterErrror: Error {
    case failedToAuthenticate
}

enum AuthenticatePresenterEvent {
    case didCancel
    case didConfirm
    case didChangePassword(text: String)
    case didChangeSalt(text: String)
}

protocol AuthenticatePresenter {
    func present()
    func handle(_ event: AuthenticatePresenterEvent)
}

final class DefaultAuthenticatePresenter {

    private weak var view: AuthenticateView?
    private let context: AuthenticateContext
    private let interactor: AuthenticateInteractor
    private let wireframe: AuthenticateWireframe

    private var password: String = ""
    private var salt: String = ""

    init(
        view: AuthenticateView,
        context: AuthenticateContext,
        interactor: AuthenticateInteractor,
        wireframe: AuthenticateWireframe
    ) {
        self.view = view
        self.context = context
        self.interactor = interactor
        self.wireframe = wireframe
    }

    func updateView() {
        view?.update(with: viewModel())
    }
}

extension DefaultAuthenticatePresenter: AuthenticatePresenter {

    func present() {
        
        updateView()
        
        guard keyStoreItem.passUnlockWithBio else { return }
        
        interactor.unlockWithBiometrics(
            keyStoreItem,
            title: Localized("authenticate.title.unlock"),
            handler: { [weak self] result in
                switch result {
                case let .success(password):
                    self?.password = password.0
                    self?.updateView()
                default:
                    ()
                }
            }
        )
        
    }

    func handle(_ event: AuthenticatePresenterEvent) {
        
        switch event {
        case let .didChangePassword(text):
            password = text
        case let .didChangeSalt(text):
            salt = text
        case .didCancel:
            wireframe.dismiss()
            context.handler(
                .failure(AuthenticatePresenterErrror.failedToAuthenticate)
            )
        case .didConfirm:
            guard interactor.isValid(
                item: keyStoreItem,
                password: password,
                salt: salt
            ) else {
                view?.animateError()
                return
            }
            wireframe.dismiss()
            context.handler(.success((password, salt)))
        }
    }
}

private extension DefaultAuthenticatePresenter {
    
    var keyStoreItem: KeyStoreItem {
        
        guard let keyStoreItem = context.keyStoreItem else {
            
            fatalError("This should never happen, the wireframe will guard against this")
        }
        
        return keyStoreItem
    }
    
    func viewModel() -> AuthenticateViewModel {

        .init(
            title: Localized("authenticate.title.unlock"),
            password: password,
            passwordPlaceholder: Localized("authenticate.placeholder.password"),
            salt: salt,
            saltPlaceholder: Localized("authenticate.placeholder.salt"),
            needsPassword: keyStoreItem.passwordType != .bio,
            needsSalt: keyStoreItem.saltMnemonic
        )
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/authenticate/AuthenticatePresenter.kt
package com.sonsofcrypto.web3walletcore.modules.authenticate

import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreItem
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreItem.PasswordType.BIO
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreItem.PasswordType.PIN
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.authenticate.AuthenticatePresenterEvent.*
import com.sonsofcrypto.web3walletcore.modules.authenticate.AuthenticateWireframeDestination.Dismiss

sealed class AuthenticatePresenterEvent {
    object DidCancel: AuthenticatePresenterEvent()
    object DidConfirm: AuthenticatePresenterEvent()
    data class DidChangePassword(val text: String): AuthenticatePresenterEvent()
    data class DidChangeSalt(val text: String): AuthenticatePresenterEvent()
}

interface AuthenticatePresenter {
    fun present()
    fun handle(event: AuthenticatePresenterEvent)
}

class DefaultAuthenticatePresenter(
    private val view: WeakRef<AuthenticateView>?,
    private val wireframe: AuthenticateWireframe,
    private val interactor: AuthenticateInteractor,
    private val context: AuthenticateWireframeContext,
): AuthenticatePresenter {
    private var password: String = ""
    private var salt: String = ""

    override fun present() {
        val keyStoreItem = keyStoreItem() ?: return
        if (interactor.canUnlockWithBio(keyStoreItem)) {
            interactor.unlockWithBiometrics(keyStoreItem, context.title) { authData, error ->
                wireframe.navigate(Dismiss)
                context.handler(authData, error)
            }
            return
        }
        updateView()
    }

    private fun keyStoreItem(): KeyStoreItem? = context.keyStoreItem ?: interactor.keyStoreItem()

    override fun handle(event: AuthenticatePresenterEvent) {
        when (event) {
            is DidCancel -> {
                wireframe.navigate(Dismiss)
                context.handler(null, null)
            }
            is DidConfirm -> {
                if (interactor.isValid(keyStoreItem()!!, password, salt)) {
                    wireframe.navigate(Dismiss)
                    context.handler(AuthenticateData(password, salt), null)
                } else {
                    view?.get()?.animateError()
                }
            }
            is DidChangePassword -> { password = event.text }
            is DidChangeSalt -> { salt = event.text }
        }
    }

    private fun updateView() = view?.get()?.update(viewModel())

    private fun viewModel(): AuthenticateViewModel =
        AuthenticateViewModel(
            Localized("authenticate.title.unlock"),
            password,
            makePassType(),
            Localized("authenticate.placeholder.password"),
            salt,
            Localized("authenticate.placeholder.salt"),
            keyStoreItem()?.passwordType != BIO,
            keyStoreItem()?.saltMnemonic ?: false
        )

    private fun makePassType(): AuthenticateViewModel.PassType =
        when (keyStoreItem()!!.passwordType) {
            PIN -> AuthenticateViewModel.PassType.PIN
            else -> AuthenticateViewModel.PassType.PASS
        }
}
